runtime/src/main/cpp/Memory.cpp

```diff
@@ -1652,4 +1669,34 @@ void MutationCheck(ObjHeader* obj) {
   if (obj->container()->frozen()) ThrowInvalidMutabilityException();
 }
 
+OBJ_GETTER(SwapRefLocked,
+    ObjHeader** location, ObjHeader* expectedValue, ObjHeader* newValue, int32_t* spinlock) {
+  lock(spinlock);
+  ObjHeader* oldValue = *location;
+  // We do not use UpdateRef() here to avoid having ReleaseRef() on return slot under lock.
+  if (oldValue == expectedValue) {
+    SetRef(location, newValue);
+  } else {
+    // We create an additional reference to the [oldValue] in the return slot.
+    if (!isArenaSlot(OBJ_RESULT) && oldValue != nullptr)
```

Comment:
Is it correct? For example, `UpdateReturnRef` has special non-trivial handling for arena slot.